lib/ocpp/v201/database_handler.cpp

```diff
@@ -616,5 +621,102 @@ OperationalStatusEnum DatabaseHandler::get_connector_availability(int32_t evse_i
     return this->get_availability(evse_id, connector_id);
 }
 
+// transactions
+void DatabaseHandler::transaction_insert(const EnhancedTransaction& transaction, int32_t evse_id) {
+    std::string sql = "INSERT INTO TRANSACTIONS (SEQ_NO, TRANSACTION_ID,EVSE_ID, CONNECTOR_ID,"
+                      "TIME_START,CHARGING_STATE,ID_TAG_SENT) VALUES"
+                      "(@seq_no, @transaction_id, @evse_id, @connector_id, @time_start, "
+                      "@charging_state, @id_token_sent)";
+    auto insert_stmt = this->database->new_statement(sql);
+
+    insert_stmt->bind_int("@seq_no", transaction.seq_no);
+    insert_stmt->bind_text("@transaction_id", transaction.transactionId.get(), SQLiteString::Transient);
+    insert_stmt->bind_int("@evse_id", evse_id);
+    insert_stmt->bind_int("@connector_id", transaction.connector_id);
+    insert_stmt->bind_datetime("@time_start", transaction.start_time);
+    insert_stmt->bind_text("@charging_state",
+                           conversions::charging_state_enum_to_string(transaction.chargingState.value()),
+                           SQLiteString::Transient);
+    insert_stmt->bind_int("@id_token_sent", transaction.id_token_sent ? 1 : 0);
+
+    if (insert_stmt->step() != SQLITE_DONE) {
+        throw QueryExecutionException(this->database->get_error_message());
+    }
+}
+
+std::unique_ptr<EnhancedTransaction> DatabaseHandler::transaction_get(const int32_t evse_id) {
+    std::string sql = "SELECT * FROM TRANSACTIONS WHERE EVSE_ID = @evse_id";
+    auto get_stmt = this->database->new_statement(sql);
+    get_stmt->bind_int("@evse_id", evse_id);
+
+    if (get_stmt->step() != SQLITE_ROW) {
+        return nullptr;
+    }
+
+    // Hardcode database_enabled to true because we would not be here otherwise
+    auto transaction = std::make_unique<EnhancedTransaction>(*this, true);
+
+    // Fill transaction
+    transaction->transactionId = get_stmt->column_text(0);
+    transaction->seq_no = get_stmt->column_int(1);
+    transaction->connector_id = get_stmt->column_int(3);
+    transaction->start_time = get_stmt->column_datetime(4);
+    transaction->chargingState = conversions::string_to_charging_state_enum(get_stmt->column_text(5));
+    transaction->id_token_sent = get_stmt->column_int(6) != 0;
+
+    if (get_stmt->step() == SQLITE_ROW) {
+        // Something strange is happening with multiple transactions
```

Comment:
This happens when there are 2 transactions in the database. It should never happen but there is no guarantee of that and this is also not the right way to solve it I would say, since you can come back from this anymore. (no way to get a transaction means no way to delete one either).

I'll take a look on how to improve this

Reply:
Should be fixed now, as well as the proper exception handling